larq/layers_base.py

```diff
@@ -3,15 +3,47 @@
 import tensorflow as tf
 
 from larq import metrics as lq_metrics
-from larq import quantizers, utils
+from larq import quantized_scope, quantized_variable, quantizers
 
 log = logging.getLogger(__name__)
 
 
 # TODO: find a good way remove duplication between QuantizerBase, QuantizerDepthwiseBase and QuantizerSeparableBase
 
 
-class QuantizerBase(tf.keras.layers.Layer):
+class BaseLayer(tf.keras.layers.Layer):
+    """Base class for defining quantized layers"""
+
+    def get_quantizer(self, name):
+        return None
+
+    def _add_variable_with_custom_getter(self, name, **kwargs):
+        quantizer = self.get_quantizer(name)
+        if quantizer is None:
+            return super()._add_variable_with_custom_getter(name, **kwargs)
+
+        old_getter = kwargs.pop("getter")
+
+        # Wrap `getter` with a version that returns a `QuantizedVariable`.
+        def getter(*args, **kwargs):
+            variable = old_getter(*args, **kwargs)
+            return quantized_variable.create_quantized_variable(variable, quantizer)
```

Comment:
What do you think of doing this as a [factory pattern](https://pythonspot.com/factory-method/)? The call would then be:

```python
return quantized_variable.QuantizedVariable.factory(variable, quantizer)
```

This may just be my Java background talking though 😅

Reply:
Hmm I opted for this since it is the way `AutoCastVariable` handles it is pretty simple. I'd be fine with something like `QuantizedVariable.from_variable` though. Since we currently don't expose the `QuantizedVariable` to the user by default, we can also change this in the future.

Reply:
OK I'm fine with that as well.